src/exchange/order.clj

```diff
@@ -0,0 +1,111 @@
+(ns exchange.order
+  (:require [clj-http.client :as http]
+            [clojure.spec.alpha :as s]
+            [next.jdbc :as jdbc]
+            [exchange.database :as database]))
+
+(def type? #{:BUY :SELL})
+(def state? #{:LIVE :FULFILLED :CANCELLED})
+
+(s/def ::id int?)
+(s/def ::user-id int?)
+(s/def ::state state?)
+(s/def ::amount int?)
+(s/def ::type type?)
+(s/def ::price int?)
+(s/def ::url string?)
+(s/def ::original-amount int?)
+(s/def ::usd-amount int?)
+(s/def ::order (s/keys :req [::user-id ::state ::amount ::type ::price ::url] :opt [::id]))
+
+(defn update-order [{::keys [id url] :as order} connection]
+  (http/post url {:form-params {:order_id id} :content-type :json})
+  (database/update-order order connection))
+
+(defn delete-order [user order-id]
+  (http/post (::url (database/get-order order-id user)) {:form-params {:order_id order-id} :content-type :json})
+  (database/delete-order order-id user))
+
+(defn get-balance-change [amount price]
+  {:SELL {:BTC (- amount) :USD (* amount price)}
+   :BUY  {:BTC amount :USD (- (* amount price))}})
+
+(defn subtract-amount [order amount price]
+  ;(println "subtract-amount" order amount)
+  (-> order
+      (update ::amount - amount)
+      (update ::state #(if (< amount (::amount order)) % :FULFILLED))
+      (update ::usd-amount + (* amount price))))
+
+(defn match-market-order! [connection {:keys [market-order, balance]} {price ::price :as standing-order}]
+  ;(println "match-market-order!" market-order balance standing-order)
+  (let [balance-limit (case (:type market-order) :SELL (:BTC balance)
+                                                 :BUY (quot (:USD balance) price))
+        trade-amount (min (:amount market-order) (::amount standing-order) balance-limit)
+        balance-change (get-balance-change trade-amount price)]
+    (update-order (subtract-amount standing-order trade-amount price) connection)
+    (database/adjust-balance connection (::user-id standing-order) (-> standing-order ::type balance-change))
+    {:market-order (-> market-order
+                       (update :amount - trade-amount)
+                       (update :USD + (* trade-amount price)))
+     :balance      (reduce #(update %1 %2 + (get-in balance-change [(:type market-order) %2])) balance [:BTC :USD])}))
+
+(defn add-market-order [user amount type]
+  (jdbc/with-transaction
+    [transaction (database/ds)]
+    (let [{{final-amount :amount usd :USD} :market-order balance :balance}
+          (reduce
+            (partial match-market-order! transaction)
```

Comment:
Is it bad practice to do side effects in reducer fn? Should I rather collect the side effects in the accumulator and execute them separately?